block/block/extractor/state.py

```diff
@@ -0,0 +1,389 @@
+        """Convert internal state map to serializable dictionary"""
+        sm_dict = dict(self._state_map)
+        for key, val in sm_dict.items():
+            sm_dict[key] = dict(val)
+            for subkey, subval in val.items():
+                sm_dict[key][subkey] = dict(subval)
+        return sm_dict
+
+    def to_msgpack(self, msgpack_path):
+        """Produce serialized blob with msgpack"""
+        with open(msgpack_path, 'wb') as file:
+            file.write(msgpack.packb((self.to_dict(), self._account_map, self._mosiac_map)))
+
+    def keys(self):
+        """Produce a view of all addresses in the state map"""
+        return self._state_map.keys()
+
+    def values(self):
+        """Produce a view of all address data in the state map"""
+        return self._state_map.values()
+
+    def items(self):
+        """Produce a list of tuples containing addresses and data"""
+        return self._state_map.items()
+
+    def insert_txn(self, txn, height, fee_multiplier):
+        # pylint: disable=too-many-locals, too-many-branches
+
+        """Insert a transaction into the state map and record resulting changes
+
+        Parameters
+        ----------
+        txn: dict
+            Deserialized transaction
+        height: int
+            Height of transaction
+        fee_multiplier: float
+            Fee multiplier for transaction's containing block
+
+        """
+
+        # need to handle flows for *all* mosaics, not just XYM
+        address = public_key_to_address(unhexlify(txn['signer_public_key']))
+
+        if txn['type'] == b'4154':  # transfer txn
+            if len(txn['payload']['message']) and txn['payload']['message'][0] == 0xfe:
+                self._state_map[address]['delegation_requests'][txn['payload']['recipient_address']].append(height)
+            elif txn['payload']['mosaics_count'] > 0:
+                for mosaic in txn['payload']['mosaics']:
+                    if hex(mosaic['mosaic_id']) in self.harvester_mosaics:
+                        self._state_map[address]['xym_balance'][height] -= mosaic['amount']
+                        self._state_map[txn['payload']['recipient_address']]['xym_balance'][height] += mosaic['amount']
+                    else:
+                        self._state_map[address]['mosaics'][hex(mosaic['mosaic_id'])].append((height, -mosaic['amount']))
+                        self._state_map[txn['payload']['recipient_address']]['mosaics'][hex(mosaic['mosaic_id'])].append(
+                            (height, mosaic['amount']))
+
+        elif txn['type'] in [b'4243', b'424c', b'414c']:  # key link txn
+            if txn['type'] == b'4243':
+                link_key = 'vrf_key_link'
+            elif txn['type'] == b'424c':
+                link_key = 'node_key_link'
+            elif txn['type'] == b'414c':
+                link_key = 'account_key_link'
+                self._account_map[public_key_to_address(txn['payload']['linked_public_key'])] = address
```

Comment:
`public_key_to_address(txn['payload']['linked_public_key'])` - extract to vat above ifs? (used here and below)